auth/couchdb_session_authenticator.go

```diff
@@ -155,49 +160,59 @@ func (a *CouchDbSessionAuthenticator) Authenticate(request *http.Request) error
 // getCookie returns an AuthSession cookie to be used in a request.
 // A new cookie will be fetched from the session end-point when needed.
 func (a *CouchDbSessionAuthenticator) getCookie() (*http.Cookie, error) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	a.flushRefreshChannel()
+
 	if a.session == nil || !a.session.isValid() {
-		err := a.syncRequestSession()
+		newSession, err := a.requestSession()
 		if err != nil {
 			return nil, err
 		}
+		a.session = newSession
 	} else if a.session.needsRefresh() {
-		ch := make(chan error)
+		// start a background process to refresh session cookie
+		// if session returned fast enough it'll be updated at the select below
+		// otherwise it'll be updated in next request at flushRefreshChannel
+		ch := make(chan error, 1)
 		go func() {
```

Comment:
What I think is still suspicious is this concurrency handling here. I need to make some search, I am not sure whether I can suggest something better than this. I will let you know when I found something.

Reply:
My best alternative suggestion is to use a `refreshOngoing bool` variable in `CouchDbSessionAuthenticator` instead of channels.
This could sign whether an requestSession already started and an other `getCoockie` can not be initiated. This way the `refresh` channel property and the `refreshTime` timestamp update in `needsRefresh()` can be removed. 
The result is simpler structure, leaves channel flush problems behind and provides the same functionality.
WDYT?

Reply:
1. `refreshTime` and your `refreshOngoing` is about controlling when to start background refresh, `refresh` channel is about message passing from background refresh to "main" instance, they are orthogonal. So no, even with adding `refreshOngoing` channel `refresh` can't be removed - they are doing different things.
2. `refreshTime` is needed because we want to start refresh ahead of session expiration to avoid `403` busts
3. I think this is needless complication for already working solution.